### ISSUE TITLE
The optional singleton pattern

### ISSUE THREAD

**Issue by arschles** [arschles/go-in-5-minutes#3]
Some packages (including some in the standard lib) do this thing where they have a bunch of top level `func`s and then a `type` (usually a struct) that has the same `func`s defined with as the receiver.

This is kinda like having a singleton class in other languages (basically a bucket of `func`s), but in this case it's optional because you can also allocate a new instance of whatever the `type` is and get the same `func`s only with the instance as the receiver. [`log`](https://godoc.org/log) has a good, clean example of what I'm talking about.

I heard Jack Lindamood name this the Optional Singleton pattern at a [GoSF meetup](http://www.meetup.com/golangsf/events/223166815/).

**Comment by gonzaloserrano** [comment]
But don't you think that in general Singletons are a bad practice because they behave like global state and because that code that uses singletons are difficult to test? e.g https://storage.googleapis.com/gtb/TotT-2008-05-15.pdf

Related with this, what do you think about a dependency injection tutorial instead?

Thanks and keep up the good work!

**Comment by arschles** [comment]
@gonzaloserrano first of all, _so awesome_ that you cited a 'testing on the toilet' paper!

anyway, I agree that in general they're a bad practice. generally I've seen this pattern applied in cases where you can also create an identical instance of the singleton if needed (e.g. for testing purposes). here's what it'd look like in a very simple logging package (this is similar to what's in the stdlib's `log` package): 

```go
package log

type Logger struct {}

func NewLogger() *Logger {
  return &Logger{}
}
func (l *Logger) Print(s string) {
  fmt.Println(s)
}

var singleton = NewLogger()

func Print(s string) {
  singleton.Print(s)
}
```

so, basically, the optional singleton pattern is for convenience, and you'd test against a `*Logger`. thoughts?

**Comment by gonzaloserrano** [comment]
Hi again,

In other languages (e.g PHP) singletons are global (e.g in PHP is implemented a static class). In the this case (and the go `log` package with the `std` var) this _singleton_ is not exported so it's just accessible  to the package, so i'd say it's not _that_ harmful (although if that variable instantiation would provoke a side-effect then testing could be trickier).

Anyway i'm not a testing expert :-)

**Comment by arschles** [comment]
you nailed it, and I think alluded to a few rules of this pattern:

- the singleton variable must be private to the package
- the singleton variable instantiation should not have side effects
- the package should use an exported function or exported type to create the singleton variable

and I'd like to propose one more:

- convenience functions that operate on the singleton variable must simply operate on the type's functions

thoughts?

**Comment by sp-gonzalo-serrano** [comment]
Sounds like a pretty good abstraction to me!

**Comment by arschles** [comment]
awesome, glad you agree. I'm going to be covering this pattern soon
